### PR TITLE
15_tictactoe - io input validation

### PR DESCRIPTION
**Pull request by riraum**
#28 
>  in io.Read, validate that the input is a valid integer
> in io.Read, validate that the input is a valid choice (i.e. not an already chosen one)

### CODE REVIEW

**Comment by riraum** [comment]
@nobe4 Tried to import `game` into `io`.
Tried to run `PlayerMove := io.Read("What's your choice? (Enter index 0-8 if not taken before)")` with a second error variable.

File: 15_tictactoe/go/io/io.go [hunk 1]
```diff
@@ -2,9 +2,25 @@ package io
...
 	var input int
-	fmt.Printf("%s > ", prompt)
-	fmt.Scan(&input)
+	for i := 0; i <= 3; i++ {
+
+		fmt.Printf("%s > ", prompt)
+		value, err := fmt.Scan(&input)
+		if err != nil {
+			continue
+		}
```
Comment (nobe4) [inline 15_tictactoe/go/io/io.go]: You should print what the error is, to help debug.

File: 15_tictactoe/go/io/io.go [hunk 2]
```diff
@@ -2,9 +2,25 @@ package io
...
+		if value == 0 || value == 1 || value == 2 || value == 3 || value == 4 || value == 5 || value == 6 || value == 7 || value == 8 {
```
Comment (nobe4) [inline 15_tictactoe/go/io/io.go]:
```suggestion
		if value >= 0 && value <= 8 {
```

File: 15_tictactoe/go/io/io.go [hunk 3]
```diff
@@ -2,9 +2,25 @@ package io
...
+			break
+		} else {
+			break
+		}
```
Comment (nobe4) [inline 15_tictactoe/go/io/io.go]: wait, so the if is 
```go
if condition {
  break
} else {
  break
}
```

Isn't that a bit useless?

File: 15_tictactoe/go/io/io.go (hunk 1 above)
Comment (nobe4) [reply 15_tictactoe/go/io/io.go]: Alternatively, you can read a string and convert the string to an int with https://pkg.go.dev/strconv

File: 15_tictactoe/go/io/io.go (hunk 3 above)
Comment (riraum) [reply 15_tictactoe/go/io/io.go]: I might have written that without too much thinking 😬
Removed the else break statement.

File: 15_tictactoe/go/io/io.go (hunk 2 above)
Comment (riraum) [reply 15_tictactoe/go/io/io.go]: That looks way more decent. Typed it like this :)

File: 15_tictactoe/go/io/io.go (hunk 1 above)
Comment (riraum) [reply 15_tictactoe/go/io/io.go]: Might have it working now.

File: 15_tictactoe/go/io/io.go (hunk 3 above)
Comment (nobe4) [reply 15_tictactoe/go/io/io.go]: I think you ought to separate the function a bit:

- have the part that does the input and loop until it's valid
- extract the validation to another function, so you can write tests for it

Comment (riraum) [reply 15_tictactoe/go/io/io.go]: Alright. Seem to also have to work on getting an error when something other than a double digit integer is entered.
That and figuring out how to check for free cells in this package.

File: 15_tictactoe/go/io/io.go [hunk 4]
```diff
@@ -2,9 +2,55 @@ package io
...
+
 func Read(prompt string) int {
 	var input int
-	fmt.Printf("%s > ", prompt)
-	fmt.Scan(&input)
+	for i := 0; i <= 3; i++ {
+
+		fmt.Printf("%s > ", prompt)
+		value, err := fmt.Scan(&input)
+		if valid(value, err) {
```
Comment (nobe4) [inline 15_tictactoe/go/io/io.go]: That's not a correct pattern in go. You ought to check the error where it's defined.

The `value` is _not_ what you typed, but the number of byte stored. You probably can discard it.

If `err` here isn't nil, then something went wrong with `fmt.Scan`, not necessarily that the value is wrong.

Comment (riraum) [reply 15_tictactoe/go/io/io.go]: I'm finding this quite hard, to separate the logic into 2 functions. I had to do it at least once before, was similarly challenging.

Did some adjustments and at least it seems to fail, with error message `out of range` when I enter 1 or several `9`s.

File: 15_tictactoe/go/io/io.go [hunk 5]
```diff
@@ -33,9 +51,11 @@ func valid(input int) bool {
 	// if value == 00 {
 	// 	return true
 	// }
+	// if the `input` is >= 0 AND <=8 return false
```
Comment (nobe4) [inline 15_tictactoe/go/io/io.go]: e.g. from our conversation, this comment adds nothing to the code below.
Try to refrain from adding noise with such comments when the code is "simple".

Comment (riraum) [reply 15_tictactoe/go/io/io.go]: Removed.

**Comment by riraum** [comment]
@nobe4 Please review!

File: 15_tictactoe/go/io/io_test.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+// 	}
+// }
+
+func TestValid(t *testing.T) {
+	tests := []struct {
+		input string
+		want  bool
+	}{
+		{"3", true},
+		{"99", false},
```
Comment (nobe4) [inline 15_tictactoe/go/io/io_test.go:13]: add a test for an empty string, for a non-number string as well

File: 15_tictactoe/go/io/io_test.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+)
+
+// func testRead(t *testing.T) {
+// 	tests := []struct {
+// 		input int
+// 		want int
+// 	}{
+// 		{}
+// 	}
+// }
```
Comment (nobe4) [inline 15_tictactoe/go/io/io_test.go]: can this be removed?

**Review by nobe4 — APPROVED** [review]
Added an improvement suggestion, otherwise good to go :shipit:

File: 15_tictactoe/go/io/io_test.go (hunk 6 above)
Comment (riraum) [reply 15_tictactoe/go/io/io_test.go:13]: Added!

File: 15_tictactoe/go/io/io_test.go (hunk 7 above)
Comment (riraum) [reply 15_tictactoe/go/io/io_test.go]: Removed!
